src/newrelic/processor.go

```diff
@@ -524,7 +526,40 @@ type harvestArgs struct {
 	blocking bool
 }
 
-func harvestPayload(p PayloadCreator, args *harvestArgs) {
+// Has all usage stats needed for the spec
+type dataUsageInfo struct {
+	endpoint_name string
+	payloadSize int
+	responseSize int
+}
+
+type dataUsageController struct {
+	duc chan dataUsageInfo
+	wg *sync.WaitGroup
+}
+
+func newDataUsageController(du_chan chan dataUsageInfo) dataUsageController {
+	return dataUsageController {
+		duc: du_chan,
+		wg: new(sync.WaitGroup),
+	}
+}
+
+func addDataUsage(duc chan dataUsageInfo, endpoint string, data_stored int, data_received int) {
+	select {
+		case duc <- dataUsageInfo{
+			         endpoint_name: endpoint,
+			         payloadSize: data_stored,
+			         responseSize: data_received,
+			        }:
+			// data stored
+		default:
```

Comment:
I think this will make the select non-blocking and could cause performance problems - see[this](https://github.com/golang/go/issues/1590) for example.

Reply:
That example hangs because the `select` is in a `for` loop, causing it to endlessly hit the `default` case. Since this is not in a loop, the non-blocking empty default will simply drop the data and move on.

Reply:
I understand  - this is a non-blocking write operation in Go.

Reply:
Added a test for this in [test(daemon): add usage metrics test for when data usage channel is f…](https://github.com/newrelic/newrelic-php-agent/pull/507/commits/cb7fdc88902f5387b30b2386f8dcc0241d70acf8)